### PR TITLE
Issue #6287 - fix classloading for WebSocketClient in webapp

### LINKED ISSUE #6287: Class loading broken for WebSocketClient used inside webapp
**Jetty version**
9.4.x

**Description**
The fix for #5859 resolved the leaking of `ClassLoader`s, but now causes class loading issues in web applications, see for example: https://github.com/cometd/cometd/issues/1037.

A web application using the WebSocket client may trigger the creation of a thread, which before #5859 was inheriting the webapp `ClassLoader` as context `ClassLoader`, but now the thread gets the server `ClassLoader` as the context `ClassLoader`.

If the task at hand needs to load a class from the webapp (for example due to JSON parsing using `oej.util.Loader.loadClass()`, it was succeeding before #5859, but fails now.

Obviously we want to keep the fix for #5859, but we need to explicitly set the context `ClassLoader` before calling webapp code (e.g. listeners, callbacks, etc.) in those components that are provided by the server, in particular `WebSocketClient`.

### PR DESCRIPTION
**Pull request by lachlan-roberts**
Issue #6287

### CODE REVIEW

File: jetty-websocket/websocket-core-client/src/main/java/org/eclipse/jetty/websocket/core/client/WebSocketCoreClient.java [hunk 1]
```diff
@@ -61,12 +62,23 @@ public WebSocketCoreClient(HttpClient httpClient, WebSocketComponents webSocketC
...
     }
 
+    public ClassLoader getClassLoader()
+    {
+        return classLoader;
+    }
+
+    public void setClassLoader(ClassLoader classLoader)
+    {
+        this.classLoader = classLoader;
```
Comment (sbordet) [inline jetty-websocket/websocket-core-client/src/main/java/org/eclipse/jetty/websocket/core/client/WebSocketCoreClient.java]: I think it should be wrapped with `Objects.requireNonNull()`, as I don't see how we ever want this class loader to be `null`.

File: jetty-websocket/websocket-javax-tests/src/test/java/org/eclipse/jetty/websocket/javax/tests/ClientClassLoaderTest.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        public void onMessage(String message)
+        {
+            textMessages.add(message);
+        }
+    }
+
+    @WebServlet("/servlet")
+    public static class WebSocketClientServlet extends HttpServlet
+    {
+        private WebSocketContainer clientContainer;
```
Comment (sbordet) [inline jetty-websocket/websocket-javax-tests/src/test/java/org/eclipse/jetty/websocket/javax/tests/ClientClassLoaderTest.java]: Allocate the WS container here and make the field `final`, as it's a more typical setup than creating a new container for every HTTP request.

**Review by sbordet — CHANGES_REQUESTED** [review]
I would like to see a similar test for Jetty's API as well.
I know the fix is in core, but we want to make sure.

File: jetty-websocket/websocket-core-client/src/main/java/org/eclipse/jetty/websocket/core/client/WebSocketCoreClient.java (hunk 1 above)
Comment (lachlan-roberts) [reply jetty-websocket/websocket-core-client/src/main/java/org/eclipse/jetty/websocket/core/client/WebSocketCoreClient.java]: I'm not even sure we really need the setter method on `WebSocketCoreClient`, it is not currently used. We always just use the ContextClassLoader at the time the `WebSocketCoreClient` was created.

But I have now changed it to use `Objects.requireNonNull()`.

**Comment by sbordet** [comment]
@lachlan-roberts apparently the tests you added for this issue fail, can you check?

**Review by sbordet — APPROVED** [review]

**Pull request merged**